File: lachancuocgoi_kotline/app/src/main/java/com/example/lachancuocgoi/Analysis/L3/core/GeminiClient.kt
package com.example.lachancuocgoi.Analysis.L3.core

import android.util.Log
import com.google.ai.client.generativeai.GenerativeModel
import com.google.ai.client.generativeai.type.generationConfig
import kotlinx.coroutines.withTimeout

/**
 * Unified Gemini API client.
 * Thay thế logic duplicate trong L3Analyzer và GeminiSummarizer.
 * 
 * Features:
 * - Automatic API key rotation khi gặp lỗi
 * - Key health tracking với cooldown đến 00:00 (quota reset)
 * - Timeout protection
 * - Generic typing để reuse cho nhiều use cases
 */
class GeminiClient(
    private val apiKeyProvider: ApiKeyProvider,
    private val config: GeminiConfig,
    private val keyHealthTracker: KeyHealthTracker? = null,
    private val onAllKeysExhausted: (() -> Unit)? = null
) {
    companion object {
        private const val TAG = "GeminiClient"
        private const val MIN_INTERVAL_MS = 1000L
        private const val CIRCUIT_BREAKER_THRESHOLD = 5
        private const val CIRCUIT_BREAKER_TIMEOUT_MS = 30_000L
    }

    private enum class CircuitState { CLOSED, OPEN, HALF_OPEN }

    @Volatile private var circuitState = CircuitState.CLOSED
    @Volatile private var consecutiveFailures = 0
    @Volatile private var circuitOpenedAt = 0L

    private var hasNotifiedAllExhausted = false

    @Volatile
    private var lastCallTimestamp = 0L

    private fun shouldAllowRequest(): Boolean {
        return when (circuitState) {
            CircuitState.CLOSED -> true
            CircuitState.HALF_OPEN -> true
            CircuitState.OPEN -> {
                val elapsed = System.currentTimeMillis() - circuitOpenedAt
                if (elapsed >= CIRCUIT_BREAKER_TIMEOUT_MS) {
                    circuitState = CircuitState.HALF_OPEN
                    Log.d(TAG, "Circuit Breaker: HALF_OPEN (testing)")
                    true
                } else {
                    false
                }
            }
        }
    }

    private fun recordSuccess() {
        consecutiveFailures = 0
        if (circuitState == CircuitState.HALF_OPEN) {
            circuitState = CircuitState.CLOSED
            Log.d(TAG, "Circuit Breaker: CLOSED (recovered)")
        }
    }

    private fun recordFailure() {
        consecutiveFailures++
        if (consecutiveFailures >= CIRCUIT_BREAKER_THRESHOLD) {
            circuitState = CircuitState.OPEN
            circuitOpenedAt = System.currentTimeMillis()
            Log.w(TAG, "Circuit Breaker: OPENED (pausing for ${CIRCUIT_BREAKER_TIMEOUT_MS}ms)")
        }
    }
    
    /**
     * Query Gemini API với automatic retry qua nhiều keys.
     * 
     * @param prompt Text prompt gửi cho AI
     * @param parser Function để parse response text thành kết quả mong muốn
     * @return Result<T> chứa parsed value hoặc error
     */
    suspend fun <T> query(
        prompt: String,
        parser: (String, String) -> T
    ): Result<T> {
        if (!shouldAllowRequest()) {
            Log.w(TAG, "Circuit Breaker: Request blocked, retry after timeout")
            return Result.failure(Exception("Circuit breaker open. Retry after ${CIRCUIT_BREAKER_TIMEOUT_MS}ms"))
        }

        // Rate limiting: Đợi nếu gọi API quá nhanh
        val now = System.currentTimeMillis()
        val elapsed = now - lastCallTimestamp
        if (elapsed < MIN_INTERVAL_MS) {
            val waitMs = MIN_INTERVAL_MS - elapsed
            Log.d(TAG, "Rate limiting: waiting ${waitMs}ms before next call")
            kotlinx.coroutines.delay(waitMs)
        }
        lastCallTimestamp = System.currentTimeMillis()
        
        val startTime = System.currentTimeMillis()
        val keys = apiKeyProvider.getApiKeys()
        
        if (keys.isEmpty()) {
            return Result.failure(Exception("No API keys available"))
        }
        
        // Lấy danh sách key ACTIVE từ KeyHealthTracker (nếu có)
        val activeIndices: List<Int> = keyHealthTracker?.getActiveKeyIndices() ?: keys.indices.toList()
        
        if (activeIndices.isEmpty()) {
            notifyAllKeysExhausted()
            return Result.failure(Exception("All API keys are in COOLDOWN/EXHAUSTED. Quota resets at 00:00."))
        }
        
        var lastException: Exception? = null
        hasNotifiedAllExhausted = false  // Reset flag cho lần gọi mới

        val fallbackModels = listOf(
            "gemini-3.1-flash-lite",
            "gemini-2.5-flash-lite",
            "gemini-3-flash",
            "gemini-2.5-flash"
        )
        
        // Thử các key ACTIVE theo thứ tự từ KeyHealthTracker
        for (keyIndex in activeIndices) {
            val apiKey = keys[keyIndex]
            var keySuccess = false
            
            for (modelName in fallbackModels) {
                try {
                    val model = createModel(apiKey, modelName)
                    val responseText = queryWithTimeout(model, prompt)
                    val parsed = parser(responseText, modelName)

                    // Success
                    keyHealthTracker?.markSuccess(keyIndex)
                    recordSuccess()
                    val latency = System.currentTimeMillis() - startTime
                    GeminiMetrics.recordCall(success = true, latencyMs = latency, keyIndex = keyIndex)
                    Log.d(TAG, "Query successful with key $keyIndex using model $modelName in ${latency}ms")
                    return Result.success(parsed)

                } catch (e: Exception) {
                    recordFailure()
                    // Xử lý chuẩn System Cancellation
                    if (e is kotlinx.coroutines.CancellationException) {
                        if (e is kotlinx.coroutines.TimeoutCancellationException) {
                            Log.w(TAG, "Timeout Exception with key $keyIndex model $modelName - Network issues")
                            keyHealthTracker?.markError(keyIndex, e.message)
                            lastException = e
                            break // Đứt mạng/timeout thì huỷ model loop để sang key kế tiếp (chắc cũng vẫn tạch) hoặc throw
                        }
                        throw e // Propagate real OS cancellations
                    }

                    Log.e(TAG, "Error with key $keyIndex model $modelName: ${e.message}")
                    lastException = e
                    
                    // Phân cực lỗi
                    val msg = e.message?.lowercase() ?: ""
                    val isQuotaError = msg.contains("429") || msg.contains("quota")
                    val isAuthError = msg.contains("403") || msg.contains("api key") || msg.contains("api_key")
                    val isModelNotFound = msg.contains("404") || msg.contains("not found")
                    
                    if (isAuthError) {
                        // Key invalid/revoked → EXHAUSTED vĩnh viễn
                        keyHealthTracker?.markInvalid(keyIndex, e.message)
                        break // Dừng model loop, sang key tiếp theo
                    } else if (isQuotaError || isModelNotFound) {
                        // Hết quota hoặc model không tồn tại trên tier này -> tiếp tục thử model khác
                        Log.w(TAG, "Quota limit or not found on model $modelName, falling back to next model...")
                        continue
                    } else {
                        // Lỗi mạng/server → mark error nhưng thử model / key khác
                        keyHealthTracker?.markError(keyIndex, e.message)
                        Log.w(TAG, "Lỗi mạng hoặc server. Dừng thử model tiếp theo trên key này.")
                        break // Dừng model loop
                    }
                }
            } // Kết thúc model loop cho key
            
            // Nếu đã thử xong tất cả model cho key này mà vẫn tạch, và lỗi cuối là liên quan đến Quota
            val finalMsg = lastException?.message?.lowercase() ?: ""
            if (finalMsg.contains("429") || finalMsg.contains("quota")) {
                keyHealthTracker?.markQuotaExceeded(keyIndex)
            }
        }
        
        // Kiểm tra xem tất cả keys có đều down không
        if (keyHealthTracker?.areAllKeysDown() == true) {
            notifyAllKeysExhausted()
        }
        
        val latency = System.currentTimeMillis() - startTime
        GeminiMetrics.recordCall(success = false, latencyMs = latency)
        return Result.failure(lastException ?: Exception("All API keys failed"))
    }
    
    /**
     * Tạo GenerativeModel instance với config.
     */
    private fun notifyAllKeysExhausted() {
        if (!hasNotifiedAllExhausted) {
            hasNotifiedAllExhausted = true
            Log.e(TAG, "ALL keys exhausted! Quota resets at 00:00.")
            onAllKeysExhausted?.invoke()
        }
    }
    
    private fun createModel(apiKey: String, modelName: String): GenerativeModel {
        val genConfig = generationConfig {
            temperature = config.temperature
            topK = config.topK
            topP = config.topP
            config.responseMimeType?.let { responseMimeType = it }
        }
        
        return GenerativeModel(
            modelName = modelName,
            apiKey = apiKey,
            generationConfig = genConfig
        )
    }
    
    /**
     * Query model với timeout protection.
     */
    private suspend fun queryWithTimeout(
        model: GenerativeModel,
        prompt: String
    ): String {
        val response = withTimeout(config.timeoutMs) {
            model.generateContent(prompt)
        }
        
        return response.text ?: throw Exception("Empty response from Gemini")
    }
}

File: lachancuoicgoi_flutter/lib/analysis/l3/core/gemini_client.dart
import 'dart:async';

import 'package:google_generative_ai/google_generative_ai.dart';

import 'api_key_provider.dart';
import 'gemini_config.dart';
import 'gemini_metrics.dart';
import 'key_health_tracker.dart';
import 'operation_result.dart';
import '../../../../core/system_logger.dart';

typedef GeminiRequestExecutor =
    Future<String> Function({
      required String apiKey,
      required GeminiConfig config,
      required String modelName,
      required String prompt,
    });

enum GeminiErrorType { quota, auth, modelNotFound, timeout, network, unknown }

/// Shared error classifier used by both [GeminiClient] and [GeminiChatSession].
GeminiErrorType classifyGeminiError(Object? error) {
  final message = error?.toString().toLowerCase() ?? '';
  if (message.contains('429') || message.contains('quota')) {
    return GeminiErrorType.quota;
  }
  if (message.contains('403') ||
      message.contains('401') ||
      message.contains('unauthorized') ||
      message.contains('api key') ||
      message.contains('api_key')) {
    return GeminiErrorType.auth;
  }
  if (message.contains('404') || message.contains('not found')) {
    return GeminiErrorType.modelNotFound;
  }
  if (message.contains('timeout')) {
    return GeminiErrorType.timeout;
  }
  if (message.contains('socket') ||
      message.contains('network') ||
      message.contains('connection')) {
    return GeminiErrorType.network;
  }
  return GeminiErrorType.unknown;
}

/// Default model priority list for Gemini API, shared across client and session.
/// The client tries each model in order, falling back to the next on failure.
/// Priority: newest/most capable first → oldest/lightest last.
const List<String> geminiFallbackModels = <String>[
  'gemini-3.5-flash',        // Ưu tiên 1: Model mới nhất, mạnh nhất
  'gemini-3.1-flash-lite',   // Ưu tiên 2: Fallback nhẹ hơn
  'gemini-3-flash',          // Ưu tiên 3: Thế hệ 3 cơ bản
  'gemini-2.5-flash',        // Ưu tiên 4: Thế hệ 2.5 đầy đủ
  'gemini-2.5-flash-lite',   // Ưu tiên 5 (cuối): Nhẹ nhất, fallback an toàn
];

class GeminiClient {
  GeminiClient({
    required this.apiKeyProvider,
    required this.config,
    this.keyHealthTracker,
    this.onAllKeysExhausted,
    GeminiRequestExecutor? requestExecutor,
  }) : _requestExecutor = requestExecutor ?? _defaultRequestExecutor;

  static const Duration _minInterval = Duration(seconds: 1);
  static const int _circuitBreakerThreshold = 5;
  static const Duration _circuitBreakerTimeout = Duration(seconds: 30);
  static const List<String> _fallbackModels = geminiFallbackModels;

  final ApiKeyProvider apiKeyProvider;
  final GeminiConfig config;
  final KeyHealthTracker? keyHealthTracker;
  final void Function()? onAllKeysExhausted;
  final GeminiRequestExecutor _requestExecutor;

  _CircuitState _circuitState = _CircuitState.closed;
  int _consecutiveFailures = 0;
  DateTime? _circuitOpenedAt;
  bool _hasNotifiedAllExhausted = false;
  // Rate-limit serialization: a Future chain that every call appends to.
  // Each call awaits the previous link, then enforces the min spacing
  // relative to the previous call's completion. This makes _applyRateLimit
  // atomic under concurrency — two overlapping query() calls can no longer
  // both read the stale timestamp and fire simultaneously (which previously
  // defeated the 1s minimum spacing).
  Future<void> _rateLimitChain = Future<void>.value();
  DateTime? _lastCallTimestamp;

  Future<Result<T>> query<T>(
    String prompt,
    T Function(String responseText, String modelName) parser,
  ) async {
    if (!_shouldAllowRequest()) {
      return Result.failure(
        StateError(
          'Circuit breaker open. Retry after ${_circuitBreakerTimeout.inMilliseconds}ms',
        ),
      );
    }

    await _applyRateLimit();
    final startTime = DateTime.now();
    final keys = apiKeyProvider.getApiKeys();
    if (keys.isEmpty) {
      return Result.failure(StateError('No API keys available'));
    }

    final activeIndices =
        keyHealthTracker?.getActiveKeyIndices() ??
        List<int>.generate(keys.length, (index) => index);
    if (activeIndices.isEmpty) {
      _notifyAllKeysExhausted();
      SystemLogger.instance.log(
        LogCategory.model,
        'Tất cả các API key của Gemini đều hết hạn mức (Quota) hoặc lỗi.',
        level: LogLevel.error,
      );
      return Result.failure(
        StateError(
          'All API keys are in COOLDOWN/EXHAUSTED. Quota resets at 00:00.',
        ),
      );
    }

    Object? lastError;
    StackTrace? lastStackTrace;

    for (final keyIndex in activeIndices) {
      final apiKey = keys[keyIndex];
      var quotaBackoffAttempt = 0;

      for (
        var modelIndex = 0;
        modelIndex < _fallbackModels.length;
        modelIndex++
      ) {
        final modelName = _fallbackModels[modelIndex];
        try {
          SystemLogger.instance.log(LogCategory.model, 'Đang gửi prompt phân tích tới Gemini ($modelName)...');
          final responseText = await _requestExecutor(
            apiKey: apiKey,
            config: config,
            modelName: modelName,
            prompt: prompt,
          );
          final parsed = parser(responseText, modelName);
          keyHealthTracker?.markSuccess(keyIndex);

          // L3 Resilience: Cost tracking
          final estimatedTokens = (prompt.length + responseText.length) ~/ 4;
          keyHealthTracker?.recordTokenUsage(keyIndex, estimatedTokens);

          _recordSuccess();
          SystemLogger.instance.log(LogCategory.model, 'Gemini phản hồi thành công ($modelName).');
          GeminiMetrics.instance.recordCall(
            success: true,
            latencyMs: DateTime.now().difference(startTime).inMilliseconds,
            keyIndex: keyIndex,
          );
          return Result.success(parsed);
        } on Object catch (error, stackTrace) {
          lastError = error;
          lastStackTrace = stackTrace;
          _recordFailure();

          final errorType = _classifyError(error);
          SystemLogger.instance.log(
            LogCategory.model,
            'Lỗi gọi Gemini ($modelName): ${error.toString().split('\n').first}. Loại: ${errorType.name}',
            level: LogLevel.warning,
          );
          if (errorType == GeminiErrorType.auth) {
            keyHealthTracker?.markInvalid(keyIndex, error.toString());
            break;
          }
          if (errorType == GeminiErrorType.quota) {
            keyHealthTracker?.markError(keyIndex, error.toString());
            if (modelIndex < _fallbackModels.length - 1) {
              quotaBackoffAttempt++;
              await Future<void>.delayed(
                Duration(seconds: quotaBackoffAttempt),
              );
              continue;
            }
            break;
          }
          if (errorType == GeminiErrorType.modelNotFound) {
            if (modelIndex < _fallbackModels.length - 1) {
              continue;
            }
            break;
          }

          keyHealthTracker?.markError(keyIndex, error.toString());
          break;
        }
      }
      final finalType = _classifyError(lastError);
      if (finalType == GeminiErrorType.quota) {
        keyHealthTracker?.markQuotaExceeded(keyIndex);
      }
    }

    if (keyHealthTracker?.areAllKeysDown() ?? false) {
      _notifyAllKeysExhausted();
    }
    GeminiMetrics.instance.recordCall(
      success: false,
      latencyMs: DateTime.now().difference(startTime).inMilliseconds,
    );
    return Result.failure(
      lastError ?? StateError('All API keys failed'),
      lastStackTrace,
    );
  }

  bool _shouldAllowRequest() {
    switch (_circuitState) {
      case _CircuitState.closed:
        return true;
      case _CircuitState.halfOpen:
        return true;
      case _CircuitState.open:
        final openedAt = _circuitOpenedAt;
        if (openedAt == null ||
            DateTime.now().difference(openedAt) >= _circuitBreakerTimeout) {
          _circuitState = _CircuitState.halfOpen;
          return true;
        }
        return false;
    }
  }

  void _recordSuccess() {
    _consecutiveFailures = 0;
    // Fix: chỉ reset cờ "all keys exhausted" khi có request thành công
    // trở lại — tránh onAllKeysExhausted bị bắn lặp ở mọi query khi
    // toàn bộ key vẫn đang down.
    _hasNotifiedAllExhausted = false;
    if (_circuitState == _CircuitState.halfOpen) {
      _circuitState = _CircuitState.closed;
    }
  }

  void _recordFailure() {
    _consecutiveFailures++;
    if (_consecutiveFailures >= _circuitBreakerThreshold) {
      _circuitState = _CircuitState.open;
      _circuitOpenedAt = DateTime.now();
    }
  }

  Future<void> _applyRateLimit() async {
    // Serialize rate-limit checks through a Future chain so concurrent callers
    // queue up rather than each reading a stale _lastCallTimestamp and firing
    // together. Each caller appends its own delay link after the previous one.
    final completer = Completer<void>();
    final previous = _rateLimitChain;
    _rateLimitChain = completer.future;
    try {
      await previous;
      final lastCall = _lastCallTimestamp;
      if (lastCall != null) {
        final elapsed = DateTime.now().difference(lastCall);
        if (elapsed < _minInterval) {
          await Future<void>.delayed(_minInterval - elapsed);
        }
      }
      _lastCallTimestamp = DateTime.now();
    } finally {
      completer.complete();
    }
  }

  void _notifyAllKeysExhausted() {
    if (_hasNotifiedAllExhausted) {
      return;
    }
    _hasNotifiedAllExhausted = true;
    onAllKeysExhausted?.call();
  }

  GeminiErrorType _classifyError(Object? error) => classifyGeminiError(error);

  static Future<String> _defaultRequestExecutor({
    required String apiKey,
    required GeminiConfig config,
    required String modelName,
    required String prompt,
  }) async {
    final model = GenerativeModel(
      model: modelName,
      apiKey: apiKey,
      generationConfig: GenerationConfig(
        temperature: config.temperature,
        topK: config.topK,
        topP: config.topP,
        responseMimeType: config.responseMimeType,
      ),
    );
    final response = await model
        .generateContent(<Content>[Content.text(prompt)])
        .timeout(config.timeout);
    final text = response.text?.trim();
    if (text == null || text.isEmpty) {
      throw StateError('Empty response from Gemini');
    }
    return text;
  }
}

enum _CircuitState { closed, open, halfOpen }
